src/evolve.ts

```diff
@@ -0,0 +1,236 @@
+import { purry } from './purry';
+import { toPairs } from './toPairs';
+
+type AFunction = (...a: Array<any>) => any;
+type Primitive = string | number | boolean | null | undefined | symbol;
+
+/**
+ * basic structure of `data` parameter and return of the function `evolve`.
+ */
+type NestedObject = Record<string, Primitive | object>;
+
+/**
+ * basic structure of `evolver` parameter of the function `evolve`.
+ */
+type EvolverStructure = Readonly<
+  Record<string, ((data: unknown) => unknown) | object | undefined>
+>;
+
+/**
+ * @example
+ * type T1 = GetValueByKey<{ A: '1', B: '2' }, 'A'>; // type T1 = "1"
+ * type T2 = GetValueByKey<{ A: '1', B: '2' }, 'B'>; // type T2 = "2"
+ * type T3 = GetValueByKey<{ A: '1', B: '2' }, 'C', '3'>; // type T3 = "3"
+ */
+type GetValueByKey<
+  T,
+  K extends PropertyKey,
+  Default = never,
+> = K extends keyof T ? T[K] : Default;
+
+/**
+ * Get the type of first parameter.
+ * But get `never` if the function take two or more required parameters.
+ * @example
+ * type Number = GetFirstParam<(arg1: number) => void>; // type Number = number
+ * type Number = GetFirstParam<(arg1: number, arg2?: number) => void>; // type Number = number
+ * type Number = GetFirstParam< (arg1: number, arg2?: number, arg3?: number) => void >; // type Number = number
+ * type Never = GetFirstParam<(arg1: number, arg2: number | undefined) => void>; // type Never = never
+ */
+type GetFirstParam<T extends AFunction> = T extends (
+  firstArg: infer Ret,
+  ...restArgs: infer Rest
+) => any
+  ? 0 extends Rest['length']
+    ? Ret
+    : never
+  : never;
+
+/**
+ * Creates an assumed `data` type from the type of the `evolver` argument.
+ * @example
+ * type Number = EvolveTarget<(arg1: number) => void>; // type Number = number
+ * type Nested = EvolveTarget<{
+ *   num: (arg1: number) => void;
+ *   obj: {
+ *    num: (arg1: number) => void
+ *   }
+ *   notFunc: undefined
+ * }>;
+ * // type Nested = {
+ * //     num?: number | undefined;
+ * //     obj?: {
+ * //         num?: number | undefined;
+ * //     } | undefined;
+ * // }
+ */
+type EvolveTarget<E> = E extends AFunction
+  ? GetFirstParam<E>
+  : E extends object
+    ? {
+        [K in keyof E as E[K] extends AFunction
+          ? K
+          : E[K] extends object
+            ? K
+            : never]?: EvolveTarget<E[K]>;
+      }
+    : never;
```

Comment:
You don't need any of this, the whole function could be implemented with just Evolver and Evolved.

```suggestion
```

Reply:
Indeed, it passed the type test without it.